modules/fedimint-meta-server/src/lib.rs

```diff
@@ -298,15 +260,60 @@ impl ServerModule for Meta {
     type Common = MetaModuleTypes;
     type Init = MetaInit;
 
+    /// Check the difference between what's desired vs submitted and consensus.
+    ///
+    /// Returns:
+    /// Items to submit as our proposal.
     async fn consensus_proposal(
         &self,
         dbtx: &mut DatabaseTransaction<'_>,
     ) -> Vec<MetaConsensusItem> {
-        let (to_submit, _to_delete) = self.desired_submitted_diff(dbtx).await;
+        let desired: Vec<_> = Self::get_desired(dbtx).await;
+
+        let mut to_submit = vec![];
+
+        for (
+            key,
+            MetaDesiredValue {
+                value: desired_value,
+                salt,
+            },
+        ) in desired
+        {
+            let consensus_value = &Self::get_consensus(dbtx, key).await;
+            let consensus_submission_value =
+                Self::get_submission(dbtx, key, self.our_peer_id).await;
+            if consensus_submission_value.as_ref()
+                == Some(&MetaSubmissionValue {
+                    value: desired_value.clone(),
+                    salt,
+                })
+            {
+                // our submission is already registered, nothing to do
+            } else if consensus_value.as_ref() == Some(&desired_value) {
+                if consensus_submission_value.is_none() {
+                    // our desired value is equal to consensus and cleared our
+                    // submission (as it is equal the
+                    // consensus) so we don't need to propose it
+                } else {
+                    // we want to submit the same value as the current consensus, usually
+                    // to clear the previous submission that did not became the consensus (we were
+                    // outvoted)
+                    to_submit.push(MetaConsensusItem {
+                        key,
+                        value: desired_value,
+                        salt,
+                    });
+                }
```

Comment:
When looking at it with @joschisan we noticed that this could be further simplified, but getting a correct version of the code was more important to me than reducing the complexity to its minimum. We can do that later since it's only the submission logic.